From a51a11bf503b0d3755a520d36ba444b32ea57ec1 Mon Sep 17 00:00:00 2001
From: Marc Mutz <marc.mutz@qt.io>
Date: Thu, 16 Oct 2025 10:33:54 +0200
Subject: [PATCH 436/553] =?UTF-8?q?QChar:=20avoid=20char{32=E2=86=9216}=5F?=
 =?UTF-8?q?t=20narrowing=20in=20(de)composeHelper()?=
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The old code used a char32_t to temporarily store a UTF-16 code
point. Upon finding it to be a surrogate, it then had to narrow it
back to char16_t for the call to QChar::surrogateToUcs4(). It also
invoked QChar(uint), a ctor which should have never existed to begin
with.

Rewrite the functions to keep the UTF-16 code point in a char16_t
variable until it's inspected to be either safe for promotion to
char32_t or else has been decoded (or decoding failed). This requires
to write 'uc = c' twice instead of once, but allows us to add a
comment on the failed-decoding path to guide the reader of the code,
improving readability.

Specifically declare `uc` as uninit'ed to provoke compiler warnings
when code is changed to forget to assign to `uc`.

Found by Clang 21's -Wcharacter-conversion.

Amends the start of the public history.

Pick-to: 6.8 6.5
Change-Id: I8d43b4c0e0b3852ddf730f349886bd6943b94f17
Reviewed-by: Ahmad Samir <a.samirh78@gmail.com>
(cherry picked from commit bda2c016f52318a52a0a6cb9fe0e6d0f21c9ddbe)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/corelib/text/qchar.cpp | 22 ++++++++++++++++------
 1 file changed, 16 insertions(+), 6 deletions(-)

diff --git a/src/corelib/text/qchar.cpp b/src/corelib/text/qchar.cpp
index 684c9fbe23d..63dd3afdb69 100644
--- a/src/corelib/text/qchar.cpp
+++ b/src/corelib/text/qchar.cpp
@@ -1839,13 +1839,18 @@ static void decomposeHelper(QString *str, bool canonical, QChar::UnicodeVersion
     const unsigned short *utf16 = reinterpret_cast<unsigned short *>(s.data());
     const unsigned short *uc = utf16 + s.size();
     while (uc != utf16 + from) {
-        char32_t ucs4 = *(--uc);
-        if (QChar(ucs4).isLowSurrogate() && uc != utf16) {
+        const char16_t c = *(--uc);
+        char32_t ucs4;
+        if (QChar::isLowSurrogate(c) && uc != utf16) {
             ushort high = *(uc - 1);
             if (QChar(high).isHighSurrogate()) {
                 --uc;
-                ucs4 = QChar::surrogateToUcs4(high, ucs4);
+                ucs4 = QChar::surrogateToUcs4(high, c);
+            } else {
+                ucs4 = c; // keep lone surrogate
             }
+        } else {
+            ucs4 = c;
         }
 
         if (QChar::unicodeVersion(ucs4) > version)
@@ -1943,13 +1948,18 @@ static void composeHelper(QString *str, QChar::UnicodeVersion version, qsizetype
     qsizetype pos = from;
     while (pos < s.size()) {
         qsizetype i = pos;
-        char32_t uc = s.at(pos).unicode();
-        if (QChar(uc).isHighSurrogate() && pos < s.size()-1) {
+        char32_t uc;
+        const char16_t c = s.at(pos).unicode();
+        if (QChar::isHighSurrogate(c) && pos < s.size() - 1) {
             ushort low = s.at(pos+1).unicode();
             if (QChar(low).isLowSurrogate()) {
-                uc = QChar::surrogateToUcs4(uc, low);
+                uc = QChar::surrogateToUcs4(c, low);
                 ++pos;
+            } else {
+                uc = c; // keep lone surrogate
             }
+        } else {
+            uc = c;
         }
 
         const QUnicodeTables::Properties *p = qGetProp(uc);
-- 
2.51.2

